caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/CadcAlmaResolver.java

```diff
@@ -0,0 +1,156 @@
+*  PURPOSE.  See the GNU Affero         PARTICULIER. Consultez la Licence
+*  General Public License for           Générale Publique GNU Affero
+*  more details.                        pour plus de détails.
+*                                       
+*  You should have received             Vous devriez avoir reçu une
+*  a copy of the GNU Affero             copie de la Licence Générale
+*  General Public License along         Publique GNU Affero avec
+*  with OpenCADC.  If not, see          OpenCADC ; si ce n’est
+*  <http://www.gnu.org/licenses/>.      pas le cas, consultez :
+*                                       <http://www.gnu.org/licenses/>.
+*
+*
+************************************************************************
+*/
+
+
+package ca.nrc.cadc.caom2.artifact.resolvers;
+
+import ca.nrc.cadc.caom2.artifact.resolvers.util.ResolverUtil;
+import ca.nrc.cadc.net.StorageResolver;
+import ca.nrc.cadc.net.Traceable;
+import ca.nrc.cadc.util.StringUtil;
+
+import java.net.URI;
+import java.net.URL;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.log4j.Logger;
+
+/**
+ * StorageResolver implementation for the ALMA archive.
+ * This class can convert an ALMA URI into a URL. The conversion is delegated to the AdResolver.
+ *
+ * @author yeunga
+ */
+public class CadcAlmaResolver implements StorageResolver, Traceable {
+    public static final String SCHEME = "alma";
+    private static final Logger log = Logger.getLogger(CadcAlmaResolver.class);
+
+    private void sanitize(String s) {
+        Pattern regex = Pattern.compile("^[a-zA-Z 0-9\\_\\.\\,\\-\\+\\@]*$");
+        Matcher matcher = regex.matcher(s);
+        if (!matcher.find()) {
+            throw new IllegalArgumentException("Invalid dataset characters.");
+        }
+    }
+   
+    private String removeNamespace(final URI uri) {
+        String archive = null;
+        String fileID = null;
+        String parts = uri.getRawSchemeSpecificPart();
+        int i = parts.indexOf('/');
+        if (i > 0) {
+            archive = parts.substring(0,i);
+            parts = parts.substring(i + 1); // namespace+fileID
+        }
+
+        i = parts.lastIndexOf('/');
+        if (i > 0) {
+            fileID = parts.substring(i + 1);
+        } else {
+            fileID = parts;
+        }
+        
+        if (!StringUtil.hasText(archive)) {
+            throw new IllegalArgumentException("cannot extract archive from " + uri);
+        }
+        
+        if (!StringUtil.hasText(fileID)) {
+            throw new IllegalArgumentException("cannot extract fileID from " + uri);
+        }
+
+        sanitize(archive);
+        sanitize(fileID);
+
+        // trim the archive to 6 characters
+        // TODO: Remove this trim when AD supports longer archive names
+        if (archive.length() > 6) {
+            archive = archive.substring(0, 6);
+        }
+        
+        return archive + "/" + fileID;
+    }
```

Comment:
Perhaps you should make this method package protected so you can write a unit test against it.

Reply:
There is already a unit test for CadcAlmaResolver. I can add more asserts, e.g. check that there is no 'namespace' in the returned URL.

Reply:
In the unit test, if we set the artifact URI to "alma:ALMA/some/name/space/bar.fits", we can expect the returned URL to end with ALMA/bar.fits. This verifies that 'some/name/space/' has been removed.